shared_model/interfaces/common_objects/amount.hpp

```diff
@@ -1,84 +1,72 @@
-#include <boost/multiprecision/cpp_int.hpp>
-#include <vector>
-
 #include "interfaces/base/model_primitive.hpp"
+
+#include <boost/multiprecision/cpp_int.hpp>
 #include "interfaces/common_objects/types.hpp"
-#include "utils/string_builder.hpp"
+#include "utils/lazy_initializer.hpp"
 
 namespace shared_model {
   namespace interface {
 
     /**
      * Representation of fixed point number
      */
-    class Amount : public ModelPrimitive<Amount> {
+    class Amount final : public ModelPrimitive<Amount> {
      public:
+      explicit Amount(const std::string &amount);
+      explicit Amount(std::string &&amount);
+
+      Amount(const Amount &o);
+      Amount(Amount &&o) noexcept;
+
       /**
        * Gets integer representation value, which ignores precision
        * @return amount represented as integer value, which ignores precision
        */
-      virtual const boost::multiprecision::uint256_t &intValue() const = 0;
+      const boost::multiprecision::uint256_t &intValue() const;
 
       /**
        * Gets the position of precision
        * @return the position of precision
        */
-      virtual types::PrecisionType precision() const = 0;
+      types::PrecisionType precision() const;
 
       /**
-       * Checks equality of objects inside
-       * @param rhs - other wrapped value
-       * @return true, if wrapped objects are same
+       * String representation.
+       * @return string representation of the asset.
        */
-      bool operator==(const ModelType &rhs) const override {
-        return intValue() == rhs.intValue() and precision() == rhs.precision();
-      }
+      std::string toStringRepr() const;
 
       /**
-       * String representation.
-       * @return string representation of the asset.
+       * Checks equality of objects inside
+       * @param rhs - other wrapped value
+       * @return true, if wrapped objects are same
        */
-      std::string toStringRepr() const {
-        if (precision() > 0) {
-          boost::multiprecision::cpp_dec_float_50 float50(intValue());
-          float50 /= pow(10, precision());
-          return float50.str(precision(), std::ios_base::fixed);
-        }
-        return intValue().str(0, std::ios_base::fixed);
-      }
+      bool operator==(const ModelType &rhs) const override;
 
       /**
        * Stringify the data.
        * @return the content of asset.
        */
-      std::string toString() const override {
-        return detail::PrettyStringBuilder()
-            .init("Amount")
-            .append("intValue", intValue().str())
-            .append("precision", std::to_string(precision()))
-            .finalize();
-      }
+      std::string toString() const override;
+
+      Amount *clone() const override;
+
+     private:
+      // lazy
+      template <typename T>
+      using Lazy = detail::LazyInitializer<T>;
+      const std::string amount_;
+
+      const Lazy<boost::multiprecision::uint256_t> multiprecision_repr_;
```

Comment:
Why lazy?

Reply:
Because it's deserialization is not that fast